src/device/net/dev_raw.rs

```diff
@@ -0,0 +1,287 @@
+use super::{MIN_BUFFER_LEN, NET_HDR_SIZE, QUEUE_RECEIVE, QUEUE_TRANSMIT, SUPPORTED_FEATURES};
+use crate::hal::Hal;
+use crate::queue::VirtQueue;
+use crate::transport::Transport;
+use crate::volatile::volread;
+use crate::{Error, Result};
+use log::{debug, info, warn};
+use zerocopy::AsBytes;
+
+/// Raw driver for a VirtIO block device.
+///
+/// This is a raw version of the VirtIONet driver. It provides non-blocking
+/// methods for transmitting and receiving raw slices, without the buffer
+/// management. For more higher-level fucntions such as receive buffer backing,
+/// see [`VirtIONet`].
+///
+/// [`VirtIONet`]: super::VirtIONet
+pub struct VirtIONetRaw<H: Hal, T: Transport, const QUEUE_SIZE: usize> {
+    transport: T,
+    mac: EthernetAddress,
+    recv_queue: VirtQueue<H, QUEUE_SIZE>,
+    send_queue: VirtQueue<H, QUEUE_SIZE>,
+}
+
+impl<H: Hal, T: Transport, const QUEUE_SIZE: usize> VirtIONetRaw<H, T, QUEUE_SIZE> {
+    /// Create a new VirtIO-Net driver.
+    pub fn new(mut transport: T) -> Result<Self> {
+        let negotiated_features = transport.begin_init(SUPPORTED_FEATURES);
+        info!("negotiated_features {:?}", negotiated_features);
+        // read configuration space
+        let config = transport.config_space::<Config>()?;
+        let mac;
+        // Safe because config points to a valid MMIO region for the config space.
+        unsafe {
+            mac = volread!(config, mac);
+            debug!(
+                "Got MAC={:02x?}, status={:?}",
+                mac,
+                volread!(config, status)
+            );
+        }
+        let send_queue = VirtQueue::new(
+            &mut transport,
+            QUEUE_TRANSMIT,
+            false,
+            negotiated_features.contains(Features::RING_EVENT_IDX),
+        )?;
+        let recv_queue = VirtQueue::new(
+            &mut transport,
+            QUEUE_RECEIVE,
+            false,
+            negotiated_features.contains(Features::RING_EVENT_IDX),
+        )?;
+
+        transport.finish_init();
+
+        Ok(VirtIONetRaw {
+            transport,
+            mac,
+            recv_queue,
+            send_queue,
+        })
+    }
+
+    /// Acknowledge interrupt.
+    pub fn ack_interrupt(&mut self) -> bool {
+        self.transport.ack_interrupt()
+    }
+
+    /// Disable interrupts.
+    pub fn disable_interrupts(&mut self) {
+        self.send_queue.set_dev_notify(false);
+        self.recv_queue.set_dev_notify(false);
+    }
+
+    /// Enable interrupts.
+    pub fn enable_interrupts(&mut self) {
+        self.send_queue.set_dev_notify(true);
+        self.recv_queue.set_dev_notify(true);
+    }
+
+    /// Get MAC address.
+    pub fn mac_address(&self) -> EthernetAddress {
+        self.mac
+    }
+
+    /// Whether can send packet.
+    pub fn can_send(&self) -> bool {
+        self.send_queue.available_desc() >= 1
+    }
+
+    /// Whether can receive packet.
+    pub fn can_recv(&self) -> bool {
```

Comment:
Maybe just remove this, `poll_receive` fills the same purpose.

Reply:
you mean `peek_used`?
Since it was used by ArceOS, so I reserve it, it has been deleted!

Reply:
I meant the `poll_receive` method that you added to `VirtIONetRaw`, which calls `peek_used` internally.

I've pushed a commit to keep `can_recv` on `VirtIONet`, by calling `VirtIONetRAW::peek_used`.